File: fs/src/main/scala/farjs/fs/popups/FolderShortcutsPopup.scala
package farjs.fs.popups

import farjs.filelist.FileListState
import farjs.filelist.stack.{WithStacks, WithStacksProps}
import farjs.fs.FSServices
import farjs.ui.popup._
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

case class FolderShortcutsPopupProps(onChangeDir: String => Unit,
                                     onClose: () => Unit)

object FolderShortcutsPopup extends FunctionComponent[FolderShortcutsPopupProps] {

  private[popups] var listPopup: ReactClass = ListPopup
  
  protected def render(compProps: Props): ReactElement = {
    val stacks = WithStacks.useStacks()
    val services = FSServices.useServices
    val (maybeItems, setItems) = useState(Option.empty[List[js.UndefOr[String]]])
    val (selected, setSelected) = useState(0)
    val props = compProps.wrapped

    val onAction: js.Function1[Int, Unit] = { index =>
      maybeItems.foreach { items =>
        items(index).foreach { dir =>
          props.onChangeDir(dir)
        }
      }
    }
    
    val onKeypress: js.Function1[String, Boolean] = { key =>
      var processed = true
      key match {
        case "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" =>
          onAction(key.toInt)
        case "-" =>
          services.folderShortcuts.delete(selected).toFuture.foreach { _ =>
            setItems(maybeItems.map(items => items.updated(selected, js.undefined)))
          }
        case "+" =>
          val stackItem = WithStacksProps.active(stacks).stack.peekLast[FileListState]()
          stackItem.state.foreach { state =>
            val dir = state.currDir.path
            services.folderShortcuts.save(selected, dir).toFuture.foreach { _ =>
              setItems(maybeItems.map(items => items.updated(selected, dir)))
            }
          }
        case _ =>
          processed = false
      }

      processed
    }

    useLayoutEffect({ () =>
      services.folderShortcuts.getAll().toFuture.map { shortcuts =>
        setItems(Some(shortcuts.toList))
      }
      ()
    }, Nil)

    maybeItems.map { items =>
      <(listPopup)(^.plain := ListPopupProps(
        title = "Folder shortcuts",
        items = js.Array(items.zipWithIndex.map { case (maybeItem, i) =>
          s"$i: ${maybeItem.getOrElse("<none>")}"
        }: _*),
        onAction = onAction,
        onClose = props.onClose,
        onSelect = setSelected,
        onKeypress = onKeypress,
        footer = "Edit: +, -"
      ))()
    }.orNull
  }
}

File: fs/popups/FolderShortcutsPopup.mjs
/**
 * @import PanelStackItem from "@farjs/filelist/stack/PanelStackItem.mjs"
 * @import { FileListState } from "@farjs/filelist/FileListState.mjs"
 */
import React, { useLayoutEffect, useState } from "react";
import ListPopup from "@farjs/ui/popup/ListPopup.mjs";
import WithStacks from "@farjs/filelist/stack/WithStacks.mjs";
import FSServices from "../FSServices.mjs";
import WithStacksProps from "@farjs/filelist/stack/WithStacksProps.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  onChangeDir(dir: string): void;
 *  onClose(): void;
 * }} FolderShortcutsPopupProps
 */

/**
 * @param {FolderShortcutsPopupProps} props
 */
const FolderShortcutsPopup = (props) => {
  const { listPopup } = FolderShortcutsPopup;

  const stacks = WithStacks.useStacks();
  const services = FSServices.useServices();
  const [maybeItems, setItems] = useState(
    /** @type {readonly (string | undefined)[] | undefined} */ (undefined)
  );
  const [selected, setSelected] = useState(0);

  /** @type {(index: number) => void} */
  function onAction(index) {
    if (maybeItems !== undefined) {
      const dir = maybeItems[index];
      if (dir !== undefined) {
        props.onChangeDir(dir);
      }
    }
  }

  /** @type {(key: string) => boolean} */
  function onKeypress(key) {
    let processed = true;
    switch (key) {
      case "0":
      case "1":
      case "2":
      case "3":
      case "4":
      case "5":
      case "6":
      case "7":
      case "8":
      case "9":
        onAction(parseInt(key));
        break;
      case "-":
        services.folderShortcuts.delete(selected).then(() => {
          if (maybeItems !== undefined) {
            const newItems = [...maybeItems];
            newItems[selected] = undefined;
            setItems(newItems);
          }
        });
        break;
      case "+":
        /** @type {PanelStackItem<FileListState>} */
        const stackItem = WithStacksProps.active(stacks).stack.peekLast();
        const state = stackItem.state;
        if (state) {
          const dir = state.currDir.path;
          services.folderShortcuts.save(selected, dir).then(() => {
            if (maybeItems !== undefined) {
              const newItems = [...maybeItems];
              newItems[selected] = dir;
              setItems(newItems);
            }
          });
        }
        break;
      default:
        processed = false;
        break;
    }
    return processed;
  }

  useLayoutEffect(() => {
    services.folderShortcuts.getAll().then((shortcuts) => {
      setItems(shortcuts);
    });
  }, []);

  return maybeItems !== undefined
    ? h(listPopup, {
        title: "Folder shortcuts",
        items: maybeItems.map(
          (maybeItem, i) => `${i}: ${maybeItem ?? "<none>"}`
        ),
        onAction: onAction,
        onClose: props.onClose,
        onSelect: setSelected,
        onKeypress: onKeypress,
        footer: "Edit: +, -",
      })
    : null;
};

FolderShortcutsPopup.displayName = "FolderShortcutsPopup";
FolderShortcutsPopup.listPopup = ListPopup;

export default FolderShortcutsPopup;
